### PR TITLE
feat: replace OtlpPipeline with exporter builders

### LINKED ISSUE #1810: OTLPExporter Pipeline issues
OTLP Exporters are configured via the helper methods offered by OTLP crate, using the idea of "pipeline", (eg: `opentelemetry_otlp::new_pipeline().logging()...`). This hides a lot of LoggerProvider details from the user, and has somewhat reasonable consistency across signals. - opentelemetry_otlp::new_pipeline().logging() vs opentelemetry_otlp::new_pipeline().tracing() vs opentelemetry_otlp::new_pipeline().metrics(), but I see some issues/challenges with that approach.

1. It is inconsistent with other exporters. For example, the examples for `stdout` show a different way of configuring the LoggerProvider, via `LoggerProvider::builder().with_simple_exporter(exporter).build()`. This does not hide `LoggerProvider` concepts from user. Most users get started with OpenTelemetry using stdout exporters, but then switching to otlp is a big shift in the way things are configured.
2. Composability is harder.. The OTLP Pipeline returns a built `LoggerProvider`, so user lose the ability to customize the `LoggerProvider` themselves. For eg: if user want to add a `RedactionProcessor`, it is not possible.
3. Exporters need to be aware of more things than it need to - One possible way to solve the above problem is to expose methods in OTLPPipeleine to allow adding processors. But then the Exporter pipeline need to always mimic the functionality additionally added in the provider. Even today, OTLPExporters does have knowledge about the runtime (tokio etc.), which it should not have.
4. The `install_simple/batch` is odd naming (of course, this is subjective!). 

The below is the way using LoggerProviderBuilder, shown in all examples except OTLP!
```
fn init_logs() -> Result<opentelemetry_sdk::logs::LoggerProvider, LogError> {
    let exporter = opentelemetry_otlp::new_exporter()
        .tonic()
        .with_endpoint("http://localhost:4317")
        .build_log_exporter()?;
    let provider: LoggerProvider = LoggerProvider::builder()
        .with_batch_exporter(exporter, runtime::Tokio)
        .with_resource(RESOURCE.clone())
        .build();

    Ok(provider)
}
```

The below uses OTLP::pipeline/helps, and shown in almost all OTLP examples.
```
fn init_logs() -> Result<opentelemetry_sdk::logs::LoggerProvider, LogError> {
    opentelemetry_otlp::new_pipeline()
        .logging()
        .with_resource(RESOURCE.clone())
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint("http://localhost:4317"),
        )
        .install_batch(runtime::Tokio)
}
```

Opening to get some feedback on these issues. I'd prefer to remove all OTLPPipeline* constructs, and make OTLP consistent with stdout exporter.

Something like:
```
let otlp_exporter = opentelemetry_otlp::LogExporter::builder::default().withTonic(TonicConfig {header : vec![...] } )...build();

let stdout_exporter = opentelemetry_stdout::LogExporterBuilder::default().with_encoder(|writer, data|Ok(serde_json::to_writer_pretty(writer, &data).unwrap())).build();

let provider: LoggerProvider = LoggerProvider::builder()
    .with_batch_exporter(exporter, runtime::Tokio)
    .with_simple_exporter(stdout_exporter)
    .with_resource(RESOURCE.clone())
    .with_log_limits(LogLimits {max_attributes : 1000})    
    .build();
```

The above is easily composable, consistent for all exporters. Exporters need not know anything about provider configurations or runtime or anything - it just does it job - exporting and accepting configuration on how to export, like which protocol or endpoint to use etc.

### PR DESCRIPTION
**Pull request by pitoniak32**
Fixes #1810

# Migration Guide

To move from `opentelemetry_otlp::new_exporter()`, and `opentelemetry_otlp::new_pipeline().{tracing,metrics,logging}()`, you will need to select the `Exporter`, and `Provider` for the signal you are using.

Below there is a guide with details about Exporters and Providers. There are also Examples for each type of signal.

## Exporter Guide
The following are the available exporters:
- SpanExporter
- MetricsExporter
- LogExporter

The exporter interface should have the same options as with the old method. For example, a tonic grpc exporter being configured old vs new:
```rust
// old
let old_exporter = opentelemetry_otlp::new_exporter()
  .tonic()
  .with_endpoint("http://localhost:4317");

// new
let new_exporter = SpanExporter::builder()
  .tonic()
  .with_endpoint("http://localhost:4317")
  .build()?; // note: you need to build the exporter after setting configuration.
```

## Provider Guide
The following are the available providers:
- TracingProvider
- SdkMeterProvider
- LoggerProvider

The provider interface should be similar to the old method. For example, a `TracerProvider` being configured old vs new:
```rust
// old
let tracer_provider: TracerProvider = opentelemetry_otlp::new_pipeline()
    .tracing()
    .with_exporter(old_exporter /* ^^ See exporter guide above ^^ */)
    .with_trace_config(Config::default().with_resource(RESOURCE.clone()))
    .install_batch(runtime::Tokio)?;

// new
let tracer_provider: TracerProvider = TracerProvider::builder()
    .with_batch_exporter(new_exporter /* ^^ See exporter guide above ^^ */, runtime::Tokio)
    .with_config(Config::default().with_resource(RESOURCE.clone()))
    .build(); // note: you need to build the provider after setting configuration.
```

# Signal Examples
### Trace Example
```rust
// old
fn init_tracer_provider() -> Result<sdktrace::TracerProvider, TraceError> {
    opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint("http://localhost:4317"),
        )
        .with_trace_config(Config::default().with_resource(RESOURCE.clone()))
        .install_batch(runtime::Tokio)
}

// new
fn init_tracer_provider() -> Result<sdktrace::TracerProvider, TraceError> {
    let exporter = SpanExporter::builder()
        .with_tonic()
        .with_endpoint("http://localhost:4317")
        .build()?;
    Ok(sdktrace::TracerProvider::builder()
        .with_config(Config::default().with_resource(RESOURCE.clone()))
        .with_batch_exporter(exporter, runtime::Tokio)
        .build())
}
```

### Metric Example
```rust
// old
fn init_metrics() -> Result<opentelemetry_sdk::metrics::SdkMeterProvider, MetricsError> {
    let export_config = ExportConfig {
        endpoint: "http://localhost:4317".to_string(),
        ..ExportConfig::default()
    };
    opentelemetry_otlp::new_pipeline()
        .metrics(runtime::Tokio)
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_export_config(export_config),
        )
}

// new
fn init_metrics() -> Result<opentelemetry_sdk::metrics::SdkMeterProvider, MetricsError> {
    let exporter = MetricsExporter::builder().with_tonic().build()?;

    let reader = PeriodicReader::builder(exporter, runtime::Tokio).build();

    Ok(SdkMeterProvider::builder()
        .with_reader(reader)
        .with_resource(RESOURCE.clone())
        .build())
}
```
### Log Example
```rust
// old
fn init_logs() -> Result<opentelemetry_sdk::logs::LoggerProvider, LogError> {
    opentelemetry_otlp::new_pipeline()
        .logging()
        .with_resource(RESOURCE.clone())
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint("http://localhost:4317"),
        )
        .install_batch(runtime::Tokio)
        .with_batch_exporter(exporter, runtime::Tokio)
        .build())
}

// new
fn init_logs() -> Result<opentelemetry_sdk::logs::LoggerProvider, LogError> {
    let exporter = LogExporter::builder()
        .with_tonic()
        .with_endpoint("http://localhost:4317")
        .build()?;
  
    Ok(LoggerProvider::builder()
        .with_resource(RESOURCE.clone())
        .with_exporter(
            exporter,
        )
        .with_batch_exporter(exporter, runtime::Tokio)
        .build())
}
```

## Changes

Started to replace the `OTLPPipeline` pattern in the `opentelemetry-otlp` crate.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by pitoniak32** [comment]
I have started looking into replicating `opentelemetry-stdout` patterns. I have gotten to a place that with very basic and defaulted behavior its similar. I just wanted to confirm that it seems reasonable before I start getting into the details of allowing more customization to the exporter builders.

File: examples/tracing-jaeger/src/main.rs
Comment (pitoniak32) [inline examples/tracing-jaeger/src/main.rs:26]: this would be an example of the refactored interface (minus the fact that currently its not capable of configuring the tonic endpoint or other things)

**Comment by pitoniak32** [comment]
my current thoughts on configuring the `tonic / http` exporter options would be something like this
```rust
let exporter = SpanExporter::builder()
    .with_tonic(
        TonicExporter::builder().with_endpoint("0.0.0.0:4317").build()?
    )
    .build()?;
```

but it feels kind of clunky to have 3 different builders, although it is more modular...

**Comment by pitoniak32** [comment]
This might be another issue / PR, but I'm noticing the `SdkMeterProvider` and `TracerProvider` have different interfaces for configuring options like `Resource` for example. I have a few ideas for possibly unifying this if its open for discussion in another location?

**Comment by cijothomas** [comment]
> This might be another issue / PR, but I'm noticing the `SdkMeterProvider` and `TracerProvider` have different interfaces for configuring options like `Resource` for example. I have a few ideas for possibly unifying this if its open for discussion in another location?

Yes, ensuring consistency across signals wherever feasible is definitely something we need to tackle soon!

File: examples/self-diagnostics/src/main.rs
Comment (cijothomas) [inline examples/self-diagnostics/src/main.rs:62]: I like this direction.

File: opentelemetry-sdk/src/metrics/exporter.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/metrics/exporter.rs:14]: lets do this in its own PR, to keep the scope for each PR very focused.

File: opentelemetry-stdout/examples/basic.rs
Comment (cijothomas) [inline opentelemetry-stdout/examples/basic.rs:31]: lets keep it simple itself. stdout is only for learning purposes. and simple exporter is also only meant for that.

File: opentelemetry-otlp/src/exporter/mod.rs
Comment (cijothomas) [inline opentelemetry-otlp/src/exporter/mod.rs:69]: also suggest to keep the renaming to a separate, small pr, so it is easier to review.

**Comment by cijothomas** [comment]
> I have started looking into replicating `opentelemetry-stdout` patterns. I have gotten to a place that with very basic and defaulted behavior its similar. I just wanted to confirm that it seems reasonable before I start getting into the details of allowing more customization to the exporter builders.
> 
> If you have a chance to review this @cijothomas I would really appreciate it!

Took a brief look and I like this direction. Left some comments, mostly about limiting the scope of PR. We really prefer short PRs each solving one particular aspect.

**Comment by pitoniak32** [comment]
> Took a brief look and I like this direction. Left some comments, mostly about limiting the scope of PR. We really prefer short PRs each solving one particular aspect.

I'll go through and remove some of the changes I made that weren't specific to this issue 👍

**Comment by pitoniak32** [comment]
As an output of this PR I would like to open a discussion / issue to talk about the following refactors:
- Renaming the types named `Metics*` to `Metric*` to follow the spec.
- Renaming `opentelemetry::metrics::Result` to `opentelemetry::metrics:MetricResult` (or doing the opposite and using the mod path for `LogResult`, and `TraceResult`).
- Modifying the `TracerProvider` to follow a similar configuration approach that is used for the other "Providers" instead of or in addition to the `with_config(...)` option for consistency across signals.

File: opentelemetry-otlp/examples/basic-otlp-http/src/main.rs
Comment (cijothomas) [inline opentelemetry-otlp/examples/basic-otlp-http/src/main.rs:42]: is it possible for endpoint, protocol to take defaults, so users does not have to specify this (unless they want nondefaults)

File: opentelemetry-otlp/examples/basic-otlp/src/main.rs
Comment (cijothomas) [inline opentelemetry-otlp/examples/basic-otlp/src/main.rs:53]: one inconsistency I notice between grpc and http is the end point - one has to append /v1/logs in http, but grpc does not require that. This is something we can address in a follow up.
I am thinking something like:
with_tonic() or with_http() is enough for builder to know which endpoint, protocol to use by default.
builder.with_tonic().build() -- this should just work with defaults
builder.with_http().build() -- this should also just work with defaults.

if user wish to change defaults, then they can use methods like with_endpoint, with_protocol etc.

Lets come back to this after merging this.

File: opentelemetry-otlp/src/exporter/http/mod.rs
Comment (cijothomas) [inline opentelemetry-otlp/src/exporter/http/mod.rs:659]: nit: reuse initial_headers variable from above

File: opentelemetry-otlp/src/logs.rs
Comment (cijothomas) [inline opentelemetry-otlp/src/logs.rs:111]: We'll need to make a detailed changelog entry as this is breaking change. I think its best to include before/after code snippets in changelog itself for some common scenario, and a link to the PR.
This PR's description can contain the full details and a mini migration guide.

**Review by cijothomas — APPROVED** [review]
LGTM. We'll need some follow ups, as noted in comments.
Most important thing is a detailed migration guide, which can be written in this PR's description for users to follow.
Also the changelog can mention the before/after for a common scenario for quick reference.

Given the significance of this change, I'll merge once we have 3 approvals:
@TommyCpp @lalitb @utpilla @hdost Please review.

**Comment by cijothomas** [comment]
> As an output of this PR I would like to open a discussion / issue to talk about the following refactors:
> 
> * Renaming the types named `Metics*` to `Metric*` to follow the spec.
> * Renaming `opentelemetry::metrics::Result` to `opentelemetry::metrics:MetricResult` (or doing the opposite and using the mod path for `LogResult`, and `TraceResult`).
> * Modifying the `TracerProvider` to follow a similar configuration approach that is used for the other "Providers" instead of or in addition to the `with_config(...)` option for consistency across signals.

:100: Very good points. Lets do it as follow ups.

**Comment by pitoniak32** [comment]
> LGTM. We'll need some follow ups, as noted in comments.
> 
> Most important thing is a detailed migration guide, which can be written in this PR's description for users to follow.
> 
> Also the changelog can mention the before/after for a common scenario for quick reference.

Awesome! I'll address comments and add the migration details / CHANGELOG entry sometime today 🙏

File: opentelemetry-otlp/examples/basic-otlp/src/main.rs
Comment (pitoniak32) [reply opentelemetry-otlp/examples/basic-otlp/src/main.rs:53]: > sounds good, after my most recent update, the `with_{tonic,http}()` is enough to determine the `http://localhost:431{7,8}` and the tests for this are updated also.
> 
> 
> 
> I will however need to look at the spec to see how the `/v1/*` is handled for the two, and I would be happy to create a new pr to get that sorted out.

EDIT: After looking back through the code, this should also already be the case, I made some changes to make it a little more clear

File: opentelemetry-otlp/src/lib.rs
Comment (lalitb) [inline opentelemetry-otlp/src/lib.rs:39]: nit - remove the `new_pipeline` reference from doc comment above (line 27)

File: opentelemetry-otlp/src/lib.rs
Comment (lalitb) [inline opentelemetry-otlp/src/lib.rs:77]: nit - not related to this PR, but I believe tonic has hard dependency on the tokio runtime. We should be changing this example to use `opentelemetry_sdk::runtime::Tokio`

File: opentelemetry-otlp/src/lib.rs
Comment (pitoniak32) [reply opentelemetry-otlp/src/lib.rs:39]: updated the doc comment to be more accurate for these changes.

File: opentelemetry-otlp/src/lib.rs
Comment (pitoniak32) [reply opentelemetry-otlp/src/lib.rs]: updated

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for the cleanup, nicely done :)

File: opentelemetry-otlp/src/exporter/http/mod.rs
Comment (utpilla) [inline opentelemetry-otlp/src/exporter/http/mod.rs:382]: Why are these new traits `HasHttpConfig` and `WithHttpConfig` needed? Could we have these methods as part of `HttpExporterBuilder like earlier`?

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: Those are so the config options are available on the Span / Metric / Log Exporter builders when with_http is used

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: If they are on the HttpExporterBuilder we would have to take that as a parameter for the with_http

Comment (utpilla) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: I don't follow. How does `with_http` influence `with_http_client`?
- `with_http` would bring in the default client (reqwest)
- `with_http_client` allows the user to bring their custom client

Both of these methods can reside in `impl HttpExporterBuilder`. What value are these new public traits adding?

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: for example, using the new traits the user gains access to the http specific config options without needing to configure a separate `HttpExporterBuilder` / `TonicExporterBuilder`. So in the following case, after `.with_http()` the two http specific methods become available, but the tonic specific ones aren't.
```rust
let exporter = SpanExporter::builder()
    .with_http()
    .with_http_client(...) // non default client like hyper
    .with_headers(...) // non default headers to send
    .build()?;

let exporter = SpanExporter::builder()
    .with_tonic()
    .with_metadata(...) // non default metadata to send
    .build()?;
```

Without the traits we would have to do something like:
```rust
let exporter = SpanExporter::builder()
    .with_http(HttpExporter::builder().with_headers(...).build())
    .build()?;
```

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: >How does `with_http` influence `with_http_client`?

`with_http()` alone will use the default values for http_client / headers, but after you say `with_http()` you can configure that http exporter with non default values like `.with_http_client(hyper)` to set a non default client before you call `build()`

Comment (cijothomas) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: To unblock this PR (its pretty big and will cause merge conflicts if delayed more), lets track this separately as a follow up ?

As per the SIG/Community call yesterday, we are delaying 0.27 release by a week, so we can do follow ups as necessary after merge. Also, I think changes to this area might be less intrusive to end users (not 100% sure).

@lalitb @utpilla Please share your thoughts on merging this as-is and follow up on these extra traits afterwards?

**Comment by cijothomas** [comment]
@pitoniak32 Very good migration guide in the PR description! We can keep editing it, as we get feedback from users.

File: opentelemetry-otlp/src/exporter/http/mod.rs
Comment (lalitb) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: Yes, this is good discussion. Traits are not affecting the API surface, so can be revisited aftewards.

Comment (utpilla) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: This PR brings a lot of new public API additions:
- `pub struct HttpConfig`
- `pub trait HasHttpConfig`
- `pub trait WithHttpConfig`
- `pub trait HasTonicConfig`
- `pub trait WithTonicConfig`
- `pub struct NoExporterBuilderSet`
- `pub struct TonicExporterBuilderSet(TonicExporterBuilder)`
- pub struct HttpExporterBuilderSet(HttpExporterBuilder)`

I would like us to achieve this with a reduce public API surface. I haven't looked at this deeply to assess whether it's feasible or not.

I also agree that even though these are public, we don't expect users to be using them. However, a reduced public API surface generally eases the maintenance burden in the long run. I would prefer that we attempt to simplify this if we are merging it now. My question would be if that simplification process demands any breaking change in the usage (especially setup of the exporter), would that be acceptable?

Comment (cijothomas) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: > My question would be if that simplification process demands any breaking change in the usage (especially setup of the exporter), would that be acceptable?

In my opinion -  Yes, given SDK/OTLP Exporter crates are still going to be alpha/beta for at-least few more weeks.

@pitoniak32 We can merge this as-is now. Would you have some bandwidth to see if this can be further simplified as a follow up?

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: Yeah I can look more into this, I won't be able to get started until tomorrow, my main desktop motherboard needs to be RMAd because of memory errors :( so I need to setup my dev env on another computer but I should still be able to look into it

Comment (pitoniak32) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: > @pitoniak32 We can merge this as-is now. Would you have some bandwidth to see if this can be further simplified as a follow up?

Should we have a discussion around what would be ideally included in the new solution? I joined the otel rust slack if you would like to chat there, or would it be better to open a discussion / issue? I can also just jump into trying alternatives and open a draft PR. But it could be good to align on what would be a nice interface to configure these exporters

Comment (utpilla) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: Thanks @pitoniak32! I'll reach out to you on Slack with my initial thoughts.

Comment (cijothomas) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: @utpilla okay to merge now?
(Trying to avoid delaying big PRs to avoid conflicts)

**Review by utpilla — APPROVED** [review]
Looks good overall from the user-facing changes.

File: opentelemetry-otlp/src/exporter/http/mod.rs
Comment (utpilla) [reply opentelemetry-otlp/src/exporter/http/mod.rs:382]: Yes.

**Comment by keltia** [comment]
Guys, I know this is still 0.x and stuff, but breaking the API on almost every release is, from a user point of view, quite painful.

**Comment by frederikhors** [comment]
Hi dear people, there is no guid for `with_batch_config()`, how to use it now with 0.27?

**Comment by frederikhors** [comment]
@pitoniak32, can you help me with `with_batch_config()`, how to use it now with 0.27?

**Comment by pitoniak32** [comment]
> @pitoniak32, can you help me with `with_batch_config()`, how to use it now with 0.27?

Sure! Do you have a specific issue? Did you checkout the migration guide in the overview comment already?

**Comment by frederikhors** [comment]
Yeah, but I cannot find `with_batch_config()` in there.

**Comment by lalitb** [comment]
@frederikhors Can you try something like (not tested)
```rust
fn init_tracer_provider() -> Result<sdktrace::TracerProvider, TraceError> {
    // Create a span exporter
    let exporter = SpanExporter::builder()
        .with_http()
        .with_protocol(Protocol::HttpBinary) 
        .with_endpoint("http://localhost:4318/v1/traces")
        .build()?;

    let batch_config = BatchConfigBuilder::default()
        .with_max_queue_size(2048)
        .build();

    let batch_processor = sdktrace::BatchSpanProcessor::builder(exporter, runtime::Tokio)
        .with_batch_config(batch_config)
        .build();

    Ok(TracerProvider::builder()
        .with_span_processor(batch_processor)
        .with_config(Config::default().with_resource(RESOURCE.clone()))
        .build())
}
```

**Comment by frederikhors** [comment]
It works.

I'm having only this error now:

```
error[E0599]: no method named `tracer` found for struct `opentelemetry_sdk::trace::TracerProvider` in the current scope
   |
87 |             .with_tracer(otlp_tracer_provider.tracer("otlp-tracer"))
   |                                               ^^^^^^
   |
  ::: C:\Users\Fred\.cargo\registry\src\index.crates.io-6f17d22bba15001f\opentelemetry-0.27.0\src\trace\tracer_provider.rs:31:8
   |
31 |     fn tracer(&self, name: impl Into<Cow<'static, str>>) -> Self::Tracer {
   |        ------ the method is available for `opentelemetry_sdk::trace::TracerProvider` here
   |
   = help: items from traits can only be used if the trait is in scope
help: trait `TracerProvider` which provides `tracer` is implemented but not in scope; perhaps you want to import it
   |
1  + use opentelemetry::trace::TracerProvider;
   |
help: there is a method `boxed_tracer` with a similar name
   |
87 |             .with_tracer(otlp_tracer_provider.boxed_tracer("otlp-tracer"))
   |                                               ~~~~~~~~~~~~
```

for the code:

```rust
let layer = tracing_opentelemetry::layer()
    .with_tracer(otlp_tracer_provider.tracer("otlp-tracer"))
    .with_filter(EnvFilter::from("INFO"))
    .boxed();
```

**Comment by lalitb** [comment]
Does importing `use opentelemetry::trace::TracerProvider as _;` helps?

**Comment by frederikhors** [comment]
> Try importing `use opentelemetry::trace::TracerProvider as _;` if that works?

It works. But ??? Is this OK?

**Comment by lalitb** [comment]
> It works. But ??? Is this OK?

Yes, this is typical of Rust — the trait must be in scope for its methods to be accessible.

**Comment by cijothomas** [comment]
Thanks all. If there are any issues/concerns with this change that need help - please open a new issue.

**Pull request merged**
